Refs #14271 - Set message for empty search results

johnpmitsch opened the pull request:
When the search results are empty, a message can be set using data-block="no-search-results-message"

waldenraines commented on app/assets/javascripts/bastion/components/nutupane.factory.js:
```diff
@@ -69,6 +69,13 @@ angular.module('Bastion.components').factory('Nutupane',
+                table.emptySearchMessage = params.emptySearchMessage || 
+                                           "No " + table.resourceName + " returned for your search query.";
+                table.emptyRowsMessage = params.emptyRowsMessage ||
+                                         "You currently don't have any " + table.resourceName;
```
These statements are untranslated.

Further more they belong in html and not in this JS file.

johnpmitsch commented:
@waldenraines updated this and https://github.com/Katello/katello/pull/6171 to your design of keeping the message in the URL. (the Katello PR is still WIP) There is one scenario that I can't seem to figure out.

When the user searches for a non-existing term, the correct "zero results found" message is displayed, but when they start backspacing, the message switches to "You don't have any _____" which is incorrect when they do have items.

I made a [screencast](https://vid.me/OQll) let me know if you need any more clarification.

Its hard to keep all of the boolean conditionals correct for every situation, do you have any ideas on how to fix this?

waldenraines commented:
> When the user searches for a non-existing term, the correct "zero results found" message is displayed, but when they start backspacing, the message switches to "You don't have any _____" which is incorrect when they do have items.

Have you checked the values of `table.searchTerm`  and `table.searchCompleted` for this case?  Has to be something incorrectly triggering the `ng-show` here - https://github.com/Katello/bastion/pull/116/files#diff-2db09d06e31eca6a952435979546fbc6R67

johnpmitsch commented:
@waldenraines thanks, I updated to fix the problem mentioned above, and also added a "refresh" button. The refresh button will clear the search and display all the results for that object. It also hides when there is no search query. I think the refresh button will communicate that the search is not only clearing the search box, but also refreshing the results. This allows users an easy way to get back to the main "full results" table.

johnpmitsch commented:
@waldenraines also want to mention that I am looking for feedback and suggestions!

waldenraines commented on app/assets/javascripts/bastion/layouts/nutupane.html:
```diff
@@ -31,6 +31,12 @@ <h2 class="col-sm-2">
+                  type="button"
+                  ng-click='table.search("") && (table.searchCompleted = false)' 
+                  ng-show="table.searchCompleted && table.searchTerm">
+            <i class="fa fa-undo"></i>
```
I personally think that an 'x' would be better here as the undo implies reload or refresh.  When I saw this I thought that clicking it would re-run the search.

waldenraines replied:
Other than that minor point this looks and functions as I would expect.

johnpmitsch commented:
@waldenraines updated the icon and kept the functionality the same

waldenraines commented:
> @waldenraines updated the icon and kept the functionality the same

Can you please open an issue for moving the 'x' into the search bar proper?  Feel free to assign to me.

johnpmitsch commented:
@waldenraines http://projects.theforeman.org/issues/15842

waldenraines commented:
ACK
